node/src/components/block_proposer.rs

```diff
@@ -364,263 +371,53 @@ impl BlockProposerReady {
     }
 
     /// Returns a list of candidates for inclusion into a block.
-    // TODO:
-    /// rename to proposed deploys
-    // TODO:
-    /// maybe use cuckoofilter
-    fn propose_deploys(
+    fn propose_proto_block(
         &mut self,
         deploy_config: DeployConfig,
         block_timestamp: Timestamp,
         past_deploys: HashSet<DeployHash>,
-    ) -> HashSet<DeployHash> {
-        // deploys_to_return = all deploys in pending that aren't in finalized blocks or
-        // proposed blocks from the set `past_blocks`
-        self.sets
+        random_bit: bool,
+    ) -> ProtoBlock {
+        let transfers = self
+            .sets
             .pending
             .iter()
-            .filter(|&(hash, deploy)| {
-                self.is_deploy_valid(deploy, block_timestamp, &deploy_config, &past_deploys)
-                    && !past_deploys.contains(hash)
+            .filter(|(_, deploy_type)| matches!(deploy_type, DeployType::Transfer(_)))
+            .take(deploy_config.block_max_transfer_count as usize);
+        let deploys = self
+            .sets
+            .pending
+            .iter()
+            .filter(|(_, deploy_type)| matches!(deploy_type, DeployType::Wasm(_)))
+            .take(deploy_config.block_max_deploy_count as usize);
+
+        let (deploys, transfers) = deploys.chain(transfers).fold(
+            (Vec::new(), Vec::new()),
+            |(mut deploys, mut transfers), (hash, deploy_type)| {
+                if self.is_deploy_valid(
+                    deploy_type.header(),
+                    block_timestamp,
+                    &deploy_config,
+                    &past_deploys,
+                ) && !past_deploys.contains(hash)
                     && !self.sets.finalized_deploys.contains_key(hash)
-            })
-            .map(|(hash, _deploy)| *hash)
-            .take(deploy_config.block_max_deploy_count as usize)
-            .collect::<HashSet<_>>()
-        // TODO: check gas and block size limits
+                {
+                    // all deploys in pending that aren't in finalized blocks or
+                    // proposed blocks from the set `past_blocks`
+                    match deploy_type {
+                        DeployType::Transfer(_) => transfers.push(*hash),
+                        DeployType::Wasm(_) => deploys.push(*hash),
+                    }
+                }
+                (deploys, transfers)
+            },
+        );
```

Comment:
This first counts the deploys and transfers, then filters, right? This means that we might end up with a block with less than maximum number of deploys and transfers even if we have more than that number of valid deploys and transfers - because there might be some invalid among the first `max` of them.

We should be filtering while iterating, then calling `take` on that filtered iterator (like before), I think.

Reply:
I agree with Bart here. The `fold` part can still filter out some of the previously chosen deploys.

Reply:
I'm still mad that he [ninja'd me by one minute](https://github.com/CasperLabs/casper-node/pull/604#discussion_r538189683), though!

Reply:
> I'm still mad that he [ninja'd me by one minute](https://github.com/CasperLabs/casper-node/pull/604#discussion_r538189683), though!

For me, it looks like you ninja'd Bart by 1 minute ;)

Reply:
Hmm… now it shows the same number of minutes for me. Must have something to do with his avatar and time dilation! :black_circle: :hole: